### ISSUE TITLE
Make vlc installation customizable

### ISSUE THREAD

**Issue by sghpjuikit** [sghpjuikit/player#15]
Provide config for selecting vlc installation. Configuration popup should be automatically open if no vlc path is set and no vlc installation in default directory is found.

**Comment by Xerus2000** [comment]
This isn't actually as easy, since for me it was only missing at development time. As of right now we would always deploy vlc with it, but ideally, to reduce the initial download size, we should create some kind of installer that asks if the user has vlc installed, and if yes points the player there or else downloads it itself.

**Comment by sghpjuikit** [comment]
I like the idea of portable "place and run" approach and generally hate any kind of installer, because you can never be sure what pollution does running the thing bring with itself. The application should never modify anything outside its directory (unless user specifically used it that way) and installer would obfuscate this policy. The download approach seems more reasonable.

It's also possible to add a configuration location for vlc and if it is empty, alert the user to do something.

**Comment by Xerus2000** [comment]
Actually this is very easy, NativeDiscovery is able to detect vlc installations on each OS by itself, if you just let it:

![image](https://user-images.githubusercontent.com/13354331/35441255-154dbdb4-02a3-11e8-9116-d66bef5de691.png)

**Comment by sghpjuikit** [comment]
Yeah, although the mechanism needs to be more robust than that and require logic on our part I think.

**Comment by Xerus2000** [comment]
since 3fb3e519e61304f2f5a62e8c43546215a2bf4bd7 the player now also detects native VLC installations

**Comment by sghpjuikit** [comment]
I have noticed that change. Good work. I'm leaving this open, because I'd like to have a read-only Config in settings showing the user the location and nature of the discovery. And perhaps, adding a custom location  discovery option.

**Comment by Xerus2000** [comment]
"showing the user the location and nature of the discovery"

I don't see the point, it's the default installation path and if there isn't one, then the user will have to select the location. Shipping with vlc is barely feasible or advisable, since it is completely different on each OS. If the user wants the player to be fully portable, he can still choose to install it inside the application directory.

**Comment by sghpjuikit** [comment]
You are ignoring the fact that the VLC installation needs to be 64-bit. I don't think 32-bit VLC works with this application.

I want to display how did the VLC get discovered and perhaps the version. Plus the custom location. That goes hand in hand.

**Comment by sghpjuikit** [comment]
@Xerus2000 You did investigate the implementation of vlc discoverers.

- Do you know if it is possible to get vlc path from the vlcj?
- Do you think it is possible to set the discoverers again at some later time?

**Comment by sghpjuikit** [comment]
After 31029c0 it is possible to do this.
- read-only vlc location config has been implemented
- it is possible to recompute MediaPlayerFactory which will rediscover vlc

Editing the vlc path config will require playback suspension and it should probably receive its own config editor to not do this on every keystroke, but instead have a 'change' button or something.

Im putting this off as a minor improvement.

**Comment by sghpjuikit** [comment]
Closing as I no longer think it is worth to implement user-definable vlc path. It would be too much hassle. The read-only config pointing to a discovered vlc instance and predefined application-relative vlc path support is all a user needs.

**Comment by Xerus2000** [comment]
> Editing the vlc path config will require playback suspension and it should probably receive its own config editor to not do this on every keystroke, but instead have a 'change' button or something.

You should have this anyways. There are more settings where it is inappropriate to instantly update the value.

> The read-only config pointing to a discovered vlc instance and predefined application-relative vlc path support is all a user needs.

I disagree. It is quite possible that the user installs his software in a different path than the default (I often do), and he should be able to use the player with that nonetheless.

**Comment by Xerus2000** [comment]
What about having a list of the directories, you can add a directory and remove ones you have added, but not edit entries, that eliminates the problem with continuous updates?

**Comment by sghpjuikit** [comment]
I agree that the use case exists, though cloning vlc is not that big of a deal right?

> There are more settings where it is inappropriate to instantly update the value.

There is support for this actually - making any value editor manual-update-only, and I am already planning refactor so it works consistently across all editors. This is actually a reason why vm options editing does not use available editors.
TLDR: there is some additional work I need to do to support this so it is actually usable.

> but not edit entries
> Playback suspension is not a problem, it's not like you change your VLC location on the fly.

Well, actually that's exactly what may happen. You remove directory where currently used vlc is in -> you suspend playback... What if you add new directory with vlc - you should resume playback, perhaps. There is an option to put in description that the settings requires app restart, but I work really hard to avoid those

**Comment by Xerus2000** [comment]
> I agree that the use case exists, though cloning vlc is not that big of a deal right?

Cloning VLC?

> Well, actually that's exactly what may happen. You remove directory where currently used vlc is in -> you suspend playback... What if you add new directory with vlc - you should resume playback, perhaps. There is an option to put in description that the settings requires app restart, but I dont know...

The behaviour in that case is really not important, because a user will only ever change this option a few times if ever, it's not like it's gonna be some annoying repeating thing. Just do what's easiest to code tbh.

**Comment by sghpjuikit** [comment]
Cloning as in copy paste the installation...

I do not think 'jsut do what is easy' is a good methodology, now or in general. It is bad UX if user adds a VLC folder and somehow needs to restart the application. User does not care about developer's convenience nor should he.

After refactoring Player.class I will know better whether this will be doable. Stay tuned.

**Comment by Xerus2000** [comment]
> Cloning as in copy paste the installation...

No. That is definitely not a good way to go, the user has a reason if they installed it somewhere else, maybe even little disk space, duplication would only amplify the problem.

> I do not think 'jsut do what is easy' is a good methodology, now or in general. It is bad UX if user adds a VLC folder and somehow needs to restart the application. User does not care about developer's convenience nor should he.

I know what you mean, but you can't do everything perfectly, you have already pushed back the release by a lot due to a few issues. And, really, who cares if adding a VLC directory is bad UX if the user only does it once? You have to weigh the costs - you could be doing more relevant things in the time you are using to polish this unimportant piece of UX.

**Comment by sghpjuikit** [comment]
Fair points. I'm not convinced, but I see your points.

**Comment by sghpjuikit** [comment]
Will be a bit more difficult to implement as File settings editor does not yet support relative paths. Can be done relatively simply using a would-be `RelativeTo(File)` constraint.

Note to self: I find working with Files very awkward. Id love to have a File interface and AbsoluteFile, AbsoluteDir, RelativeFile, RelativeDir classes. This would allow some neat optimizations and guarantees.

**Comment by sghpjuikit** [comment]
The Player refactor is going ok. Already have it in Kotlin (now starts the hard part).

So I decided to give this a shot and I plan on making JavaFX playback available as well. Unless some problem pops up (like a bug).

**Comment by sghpjuikit** [comment]
Done in 67cf36e.
Change in settings requires app restart as stated in config description. I also updated installation documentation.

**Comment by Xerus2000** [comment]
> If the location is child of application location, it is saved as relative.

And if the location is relative, but not a child of application location? Doesn't it just save what you enter? Haven't tested it, but this doesn't sound like expected behaviour.

**Comment by sghpjuikit** [comment]
> And if the location is relative...

Relative path must be relative to an absolute one, and there can never be more than one, since file can not have multiple parents and its parent (that can be dynamic) semantically represents one thing. If multiple files were relative to multiple parents then how would you decide which? They would no longer be relative, as the information would have to be in their path. This answers why there is always 1 relative path resolution target. And in VLC location case, this is application location.

Btw, File editor does not support manual edit right now (only using file picker), due to design issue that relates to config editors in general and which I need to take care of first.

But, if manual edit was allowed, by default file editor would only accept absolute paths. If a `Constraint.RelativeTo(target: File)` is used (like for VLC location), all files within that would be relative (whether user supplied it as relative or not). This is only natural, relative file means that some parent can change, thus all files within that parent must always be stored relative. This explains why user has no say in this and how he chooses/writes the path is merely UX.

**Comment by Xerus2000** [comment]
No, I was not talking about multiple parents. I meant if for example the vlc directory is parallel to the player directory, then the path would be `..\vlc`, which is relative to the player, but not a child of it. 

Not supporting manual editing is not a good thing, because 
1) maybe you'd like to just paste in the path instead of navigating to it 
2) user can't decide whether it should be relative or not if he'd like to do so 
3) graphical file-choosers can fail ;)

Using variables like the user home in paths does sound nice tho.

**Comment by sghpjuikit** [comment]
I see. I have sort of overlooked that case. The problem is that going 'up' is sort of going outside of one's responsibility. Manual edit will support this, file picker will not, both intentionally.

Manual edit in file editor will be supported, just needs some prior work.

The placeholder idea was not supposed to be able to refer to environment variables, just be a placeholder of predefined parent target, specified in the description of the editor. How would the editor resolve the placeholder into a file would be defined by developer for each config. So it could satisfy your idea as well.

**Comment by Xerus2000** [comment]
> How would the editor resolve the placeholder into a file would be defined by developer for each config. 

I'd rather have it be consistent across the whole application. If you just take something like environment variables, it's still easy for the developer to add custom values.

**Comment by sghpjuikit** [comment]
What do you mean consistent? Why would the placeholders be tied to environment variables or anything?

See this:
```
@Config("Vlc locations", description = "Vlc locations. For relative files, use %{abc} for program files directory and ${app} for application directory.")
val vlcLocations by cList(initialLocations...)
    .only(DIRECTORY)
    .relativeTo( // takes list of map entries of type (placeholder -> lazy file)
        "abc" to { obtainProgramFilesDir() },
        "app" to { APP.location }
    )
```

The placeholder is defined with configurable value and is bound to it and has nothing to do with any other configurable value or any environment variable or anything. If programmer obtains the value by reading `System.properties` or whatever, that is his choice and he is free to document this in the description for the user. If programmer names the placeholder based on some convention or system properties, that's his choice too (but I'd try to talk him out of it, because it is the how, not what and should be merely mentioned in description).

This approach is highly flexible and puts the burden of responsibility on the dev. If you start binding the placeholders to env. variables, how do you make sure they they work across platform and version?

Having predefined set of placeholders, even just for convenience seems like a very bad idea to me. Again, this can be done with the above idea, internally. If it makes sense, like certain values looking up the same (sub)set of locations or something.

In any case, the placeholders being tied to the configurable value is very powerful and transparent and is a good fit, because the value gives the placeholders semantics, not vice versa. And this semantics must not escape beyond the value's control. Reuse is just reuse, exactly the same as composition over inheritance (with the latter being inherently wrong way about it, because the responsibilities are dictated the wrong direction).

My 2 cents. Never thought about relative file support before now.
